### PR TITLE
Http filter dynamic forward proxy: bug fix (attempt two)

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
This PR is basically the below PR adding a unit test:
https://github.com/envoyproxy/envoy/pull/29764

For details of the problem and the fix, please check the above PR.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/common/dynamic_forward_proxy/dns_cache.h [hunk 1]
```diff
@@ -45,6 +45,8 @@ class DnsHostInfo {
    */
   virtual Network::Address::InstanceConstSharedPtr address() const PURE;
 
+  virtual bool firstResolveComplete() const PURE;
```
Comment (alyssawilk) [inline source/extensions/common/dynamic_forward_proxy/dns_cache.h:48]: mind commenting?

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 2]
```diff
@@ -408,6 +408,7 @@ void DnsCacheImpl::finishResolve(const std::string& host,
     primary_host_info->host_info_->setAddresses(new_address, std::move(address_list));
 
     runAddUpdateCallbacks(host, primary_host_info->host_info_);
+    primary_host_info->host_info_->setFirstResolveComplete();
```
Comment (alyssawilk) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:411]: do you mind runtime guarding this just out of abject paranoia?

**Review by alyssawilk — COMMENTED** [review]
awesome, LGTM modulo two nits

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc (hunk 2 above)
Comment (yanjunxiang-google) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:411]: ACK

Comment (yanjunxiang-google) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc]: done

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**